### PR TITLE
Fix `FolderBrowserDialog.ShowDialog`

### LINKED ISSUE #7981: The original FolderBrowserDialog cannot be opened
### .NET version

8.0.100-alpha.1.22514.14 

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

yes, cannot repro in .Net 7.0 

### Issue description

The FolderBrowserDialog cannot be opened when setting AutoUpgradeEnabled as false. This issue cannot be resolved after the PR fixing: https://github.com/dotnet/winforms/pull/7978
**.Net 8.0 result:**
image
_System.AccessViolationException
  HResult=0x80004003
  Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
  Source=Cannot evaluate the exception source
  StackTrace:
Cannot evaluate the exception stack trace_

**.Net 7.0 result:**
image

### Steps to reproduce

1. Open the [TestApp.zip](https://github.com/dotnet/winforms/files/9835559/TestApp.zip) project in VS
2. Build and run it
3. Click the button in the form and observe the result

### PR DESCRIPTION
**Pull request by elachlan**
Fixes #7981

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7982)

### CODE REVIEW

**Comment by elachlan** [comment]
CC: @lonitra @JeremyKuhne

**Comment by RussKie** [comment]
We can probably enumerate open windows and search for the dialog window.
In UIIntegration project we have methods to search for controls, maybe something like that can be used...
Just random thoughts

________________________________

File: src/System.Windows.Forms/src/System/Windows/Forms/FolderBrowserDialog.cs [hunk 1]
```diff
@@ -491,11 +490,15 @@ private static unsafe int FolderBrowserDialog_BrowseCallbackProc(HWND hwnd, uint
                 case PInvoke.BFFM_SELCHANGED:
                     // Indicates the selection has changed. The lpData parameter points to the item identifier list for the newly selected item.
                     ITEMIDLIST* selectedPidl = (ITEMIDLIST*)lParam;
-                    if (selectedPidl is not null)
+                    Span<char> buffer = stackalloc char[PInvoke.MAX_PATH + 1];
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/FolderBrowserDialog.cs:493]: This should be inside the `if (selectedPidl is not null)` and should assign to `char* buffer` so you can just pass it as the argument. Then you can avoid the unnecessary pin syntax.

**Review by JeremyKuhne — CHANGES_REQUESTED** [review]
Thanks for looking into this! Just one comment.

**Comment by elachlan** [comment]
The other issue was we were passing lParam to `GCHanle.FromIntPtr` which caused another exception since its `null` when `msg == 'BFFM_INITIALIZED'`.

**Comment by elachlan** [comment]
With regards to the tests. I looked at adding a `RunCommonDialogAsync` to `ControlTestBase` Or making a `ComponentTestBase`. Issue is that `CommonDialog` doesn't have a `Close()` method, since its always called via `ShowDialog`. I think someone more skilled will need to put together a plan so we can test the folder/save/open dialogs.

**Comment by RussKie** [comment]
> I have no idea how to write a test for this since `showdialog()` will block.

Could probably try something like this: https://stackoverflow.com/a/42405020/2338036.

**Comment by elachlan** [comment]
> > I have no idea how to write a test for this since `showdialog()` will block.
> 
> Could probably try something like this: https://stackoverflow.com/a/42405020/2338036.

Thanks, So I'd have to call `ShowDialog` in a different thread and then call to window via that code?

**Comment by RussKie** [comment]
I think - the other way around. Here's a raw braindump of how I'd try it - spawn a new thread (e.g., a Windows Forms timer) that would search for a window and cap the lifetime of that thread too (e.g., give it like 3 seconds, else fail with timeout). If the window is found - close it - send WM_DESTROY message or ESC key...

**Review by JeremyKuhne — APPROVED** [review]
Thanks! Going to put this in to get manual coverage of the scenario. You can follow up with a unit test for it on another PR.

**Comment by RussKie** [comment]
@JeremyKuhne we shouldn't be merging fixes without tests, as then it makes it harder to validate. There was nothing pressing about this regression and the test was already being developed.

**Pull request merged**